Doc/extending/first-extension-module.rst

```diff
@@ -0,0 +1,569 @@
+your extension.
+This should fail with the following exception:
+
+.. code-block:: pycon
+
+   >>> import spam
+   Traceback (most recent call last):
+      ...
+   ImportError: dynamic module does not define module export function (PyModExport_spam or PyInit_spam)
+
+
+Including the Header
+====================
+
+Now, add the first line to your file: include :file:`Python.h` to pull in
+all declarations of the Python C API:
+
+.. literalinclude:: ../includes/capi-extension/spammodule-01.c
+   :start-at: <Python.h>
+   :end-at: <Python.h>
+
+Next, include the header for the :c:func:`system` function:
+
+.. literalinclude:: ../includes/capi-extension/spammodule-01.c
+   :start-at: <stdlib.h>
+   :end-at: <stdlib.h>
+
+Be sure to put this, and any other standard library includes, *after*
+:file:`Python.h`.
+On some systems, Python may define some pre-processor definitions
+that affect the standard headers.
+
+.. tip::
+
+   The ``<stdlib.h>`` include is technically not necessary.
+   :file:`Python.h` :ref:`includes several standard header files <capi-system-includes>`
+   for its own use and for backwards compatibility,
+   and ``stdlib`` is one of them.
+   However, it is good practice to explicitly include what you need.
+
+With the includes in place, compile and import the extension again.
+You should get the same exception as with the empty file.
+
+.. note::
+
+   Third-party build tools should handle pointing the compiler to
+   the CPython headers and libraries, and setting appropriate options.
+
+   If you are running the compiler directly, you will need to do this yourself.
+   If your installation of Python comes with a corresponding ``python-config``
+   command, you can run something like:
+
+   .. code-block:: shell
+
+      gcc --shared $(python-config --cflags --ldflags) spammodule.c -o spam.so
+
+
+Module export hook
+==================
+
+The exception you got when you tried to import the module told you that Python
+is looking for a "module export function", also known as a
+:ref:`module export hook <extension-export-hook>`.
+Let's define one.
+
+First, add a prototype below the ``#include`` lines:
+
+.. literalinclude:: ../includes/capi-extension/spammodule-01.c
+   :start-after: /// Export hook prototype
+   :end-before: ///
+
+.. tip::
+   The prototype is not strictly necessary, but some modern compilers emit
+   warnings without it.
+   It's generally better to add the prototype than to disable the warning.
+
+The :c:macro:`PyMODEXPORT_FUNC` macro declares the function's
+return type, and adds any special linkage declarations needed
+to make the function visible and usable when CPython loads it.
+
+After the prototype, add the function itself.
+For now, make it return ``NULL``:
+
+.. code-block:: c
+
+   PyMODEXPORT_FUNC
+   PyModExport_spam(void)
+   {
+      return NULL;
+   }
+
+Compile and load the module again.
+You should get a different error this time.
```

Comment:
Should this say to restart the Python interpreter?
(I don’t know if the previous non-loading module will be cached in sys.modules)

Reply:
A non-loading one won't be cached, but I added a note to the later sections where the module loads.
Thanks!